[ Enhancement ] Improve OpenAI thread management efficiency

jerome-white opened the pull request:
## Summary

Target issue is #83 

OpenAI interaction in the platform is less performant than directly in Glific. This PR updates the code with suggestions from the issue thread.

## Checklist

Before submitting a pull request, please ensure that you mark these task.

- [ ] Ran `fastapi run --reload app/main.py` or `docker compose up` in the repository root and test.
- [ ] If you've fixed a bug or added code that is tested and has test cases.

## Notes

Reviewers feel free to push other commits -- this can be a shared review

---

This pull request aims to enhance the efficiency of OpenAI thread management within the `ProjectTech4DevAI/ai-platform` repository. The changes are made in the `backend/app/api/routes/threads.py` file and include the introduction of a `RequestCopier` class, improvements in error handling for OpenAI API calls, and a restructuring of callback handling. The source branch for these enhancements is `enhancement/open-ai-threads`, and the target branch is `main`.

jerome-white committed:
Simplified import structuring

jerome-white committed:
If thread is provided assume it is valid

jerome-white committed:
Clean up thread handling logic

* Use class variables to manage resource allocation
* Focused try-except blocks
* Functions for common needs

jerome-white committed:
Fix errors identified by Kodus

AkhileshNegi commented on backend/app/api/routes/threads.py:
```diff
@@ -22,62 +37,63 @@ def send_callback(callback_url: str, data: dict):
     This function is run in the background after we have already returned an initial response.
     """
+
+    thread_id = request["thread_id"]
```
need validation as current code breaks, as mentioned in the comment

jerome-white replied:
What do you mean by "validation"? Is it possible that this key may not appear in the request?

jerome-white committed:
Remove trailing whitespace

jerome-white committed:
Corrected variable reference

jerome-white committed:
Corrected variable management

jerome-white committed:
Start the background process sooner

Do not make the user pay for Python dictionary allocation and function
call overhead; which is what is potentially happening when this is
called after the `initial_response` variable is created.

jerome-white committed:
Merge branch 'main' into enhancement/open-ai-threads
